components/script/dom/bindings/structuredclone.rs

```diff
@@ -241,112 +277,144 @@ static STRUCTURED_CLONE_CALLBACKS: JSStructuredCloneCallbacks = JSStructuredClon
     canTransfer: Some(can_transfer_callback),
 };
 
-struct StructuredCloneHolder {
-    blob: Option<DomRoot<Blob>>,
+/// A data holder for results from, and inputs to, structured-data read/write operations.
+/// https://html.spec.whatwg.org/multipage/#safe-passing-of-structured-data
+pub enum StructuredDataHolder {
+    Read {
+        /// A deserialized blob, stored temporarily here to keep it rooted.
+        blob: Option<DomRoot<Blob>>,
+        /// A vec of transfer-received DOM ports,
+        /// to be made available to script through a message event.
+        message_ports: Option<Vec<DomRoot<MessagePort>>>,
+        /// A map of port implementations,
+        /// used as part of the "transfer-receiving" steps of ports,
+        /// to produce the DOM ports stored in `message_ports` above.
+        port_impls: Option<HashMap<MessagePortId, MessagePortImpl>>,
+    },
+    /// A data holder into which transferred ports
+    /// can be written as part of their transfer steps.
+    Write(Option<HashMap<MessagePortId, MessagePortImpl>>),
 }
 
-/// A buffer for a structured clone.
-pub enum StructuredCloneData {
-    /// A non-serializable (default) variant
-    Struct(*mut u64, size_t),
-    /// A variant that can be serialized
-    Vector(Vec<u8>),
-}
+// TODO: should this be unsafe?
```

Comment:
If it accepts a script_runtime::JSContext instead of a raw pointer then it should be ok.